# Task 4.8.1

Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [5]:
# Turn project folder path into a string
path = r'C:\Users\marze\CF Projects\08-2024 Instacart Basket Analysis'

In [10]:
# Import orders_products_merged.pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new_columns.pkl'))

In [12]:
#check dimensions
ords_prods_merge.shape

(32404859, 19)

# Task 4.8.2

In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [16]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

# Task 4.8.3

Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

The mean for each department_id has changed. Some values slightly increased while the others decreased. Some of the examples are department_id '1' is 15.577493 in the subset while 15.457838 in the whole dataframe. Also department_id 6 was 16.658449 for the subset and 16.439806 for the whole dataframe. 

# Task 4.8.4

Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [31]:
# groupby and transform() functions with np.max argument to determine maximum order number for each user_id
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\marze\AppData\Local\Temp\ipykernel_9008\1365695282.py:2: FutureWarning: The provided callable <function max at 0x000002D579909EE0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [33]:
# View updated data frame
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                      3.0        False                  5          0   both   
1                     20.0        False                  1          1   both   
2                      6.0        False                 20          0   both   
3                      NaN         True                 10          0   both   
4                      9.0        False                 11          1   both   
5                     26.0        False                  7          0   both   
6                     30.0        False                  2          0   both   
7                     30.0        False                  1          0   both   
8                     19.0        False                  1          0   both   
9                     15.0        False                  2          1   both   
10                     NaN         True                  1          0   both   
11                     7.0        False                  1          0   both   
12                     7.0        False                  3          1   both   
13                    12.0        False                  4          1   both   
14                     2.0        False                  8          0   both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy           Most orders   
1   Mid-range product  Regularly busy  Regularly bu

In [39]:
# use loc function for loyalty_flag column with given criteria
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [41]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [43]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [45]:
# Check frequency of values
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [47]:
# return the first 60 rows of only the “user_id,” “loyalty_flag,” and “order_number” columns
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

# Task 4.8.5

The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [55]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'median']})

prices                     
                       mean  min      max median
loyalty_flag                                    
Loyal customer    10.386336  1.0  99999.0    7.4
New customer      13.294670  1.0  99999.0    7.4
Regular customer  12.495717  1.0  99999.0    7.4

# Task 4.8.6

The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [61]:
# new column for mean price
ords_prods_merge['mean_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [63]:
# check columns
ords_prods_merge[['user_id', 'mean_price']].head(20)

user_id  mean_price
0       138    6.935811
1       138    6.935811
2       709    7.930208
3       764    4.972414
4       764    4.972414
5       777    6.935398
6       825    5.957576
7       910    6.680000
8      1052    7.162500
9      1052    7.162500
10     1379    5.813636
11     1480    5.081633
12     1480    5.081633
13     1480    5.081633
14     1494    6.803797
15     1494    6.803797
16     1494    6.803797
17     1540    6.607122
18     1540    6.607122
19     1540    6.607122

In [65]:
# assign flag for "low spender"
ords_prods_merge.loc[ords_prods_merge['mean_price'] < 10, 'spending_flag'] = 'Low spender'

In [67]:
# assign flag for "high spender"
ords_prods_merge.loc[ords_prods_merge['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [69]:
# check frequency of new spending_flag column
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [75]:
# check columns
ords_prods_merge[['user_id', 'mean_price', 'spending_flag']].head(100)

user_id  mean_price spending_flag
0       138    6.935811   Low spender
1       138    6.935811   Low spender
2       709    7.930208   Low spender
3       764    4.972414   Low spender
4       764    4.972414   Low spender
..      ...         ...           ...
95    10831    7.494286   Low spender
96    11119    5.878641   Low spender
97    11186    6.618085   Low spender
98    11243    5.927149   Low spender
99    11448    5.981250   Low spender

[100 rows x 3 columns]

# Task 4.8.7

In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows: 
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer. 
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer ”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [82]:
# new column for median days between orders
ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [86]:
# check columns
ords_prods_merge[['user_id', 'days_since_prior_order','median_days_between_orders']].head(20)

user_id  days_since_prior_order  median_days_between_orders
0       138                     3.0                         8.0
1       138                    20.0                         8.0
2       709                     6.0                         8.0
3       764                     NaN                         9.0
4       764                     9.0                         9.0
5       777                    26.0                        11.0
6       825                    30.0                        20.0
7       910                    30.0                         6.0
8      1052                    19.0                        10.0
9      1052                    15.0                        10.0
10     1379                     NaN                        20.0
11     1480                     7.0                         7.0
12     1480                     7.0                         7.0
13     1480                    12.0                         7.0
14     1494                     2.0                         7.0
15     1494                     6.0                         7.0
16     1494                    10.0                         7.0
17     1540                     7.0                         7.0
18     1540                     6.0                         7.0
19     1540                    14.0                         7.0

In [88]:
# assign flag for "non-frequent customer"
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [90]:
# assign flag for "regular customer"
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] <= 20) & (ords_prods_merge['median_days_between_orders'] > 10),'frequency_flag'] = 'Regular customer'

In [92]:
# assign flag for "frequent customer"
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [94]:
# check frequency of new frequency_flag column
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [96]:
# check columns
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'frequency_flag']].head(100)

user_id  days_since_prior_order  median_days_between_orders  \
0       138                     3.0                         8.0   
1       138                    20.0                         8.0   
2       709                     6.0                         8.0   
3       764                     NaN                         9.0   
4       764                     9.0                         9.0   
..      ...                     ...                         ...   
95    10831                    23.0                        23.0   
96    11119                    30.0                        13.0   
97    11186                     2.0                         6.0   
98    11243                     0.0                         6.0   
99    11448                     3.0                        14.0   

           frequency_flag  
0       Frequent customer  
1       Frequent customer  
2       Frequent customer  
3       Frequent customer  
4       Frequent customer  
..                    ...  
95  Non-frequent customer  
96       Regular customer  
97      Frequent customer  
98      Frequent customer  
99       Regular customer  

[100 rows x 4 columns]

# Task 4.8.9 

Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [100]:
# Check shape before exporting
ords_prods_merge.shape

(32404859, 25)

In [102]:
# Export data frame as pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_grouped_aggregated.pkl'))